src/lib/ffi/ffi_pkey_algs.cpp

```diff
@@ -333,6 +333,24 @@ int botan_pubkey_rsa_get_n(botan_mp_t n, botan_pubkey_t key)
    }
 
 /* DSA specific operations */
+int botan_privkey_create_dsa(botan_privkey_t* key, botan_rng_t rng_obj, size_t pbits, size_t qbits)
+   {
+#if defined(BOTAN_HAS_DSA)
+
+    if(rng_obj == nullptr)
+      return BOTAN_FFI_ERROR_NULL_POINTER;
+
+    return ffi_guard_thunk(BOTAN_CURRENT_FUNCTION, [=]() -> int {
+      Botan::RandomNumberGenerator& rng = safe_get(rng_obj);
+      Botan::DL_Group group(rng, Botan::DL_Group::Prime_Subgroup, pbits, qbits);
```

Comment:
For DSA this should probably instead use the NIST approved DSA parameter generator (DL_Group::DSA_Kosherizer)

Reply:
Also, otherwise it is possible to generate a DSA key with params that are considered invalid for DSA (eg a 1536 bit p with 384 bit q). Such params are fine, technically speaking, but some implementations may reject such keys outright for not following the NIST param lengths (1024+160, 2048+256, etc)

Reply:
I'm not using DSA_Kosherizer because in OpenPGP we want to be able to generate any key between 1024 and 3074 with size that's multiple of 64.